File: pcli/pyapi/pyAPI.cxx
#include <boost/python.hpp>
#include <boost/python/return_internal_reference.hpp>
#include "pyAPI.hxx"


// static pHost pLocalHost("localhost" );


#ifdef NOPYTHON

main()
{
    pHost localhost("localhost");
    pFile f( "a.out", localhost );
}

#else
using namespace boost::python;

// template <class T>
// struct describe
// {
//     static PyObject* convert(const T& x)
//     {
//         return boost::python::incref(x.describe().ptr());
//     }
// };

list listPidsOnHost( pHost host )
{
    return pFrameWork::listPidsOnHost( host );
}


BOOST_PYTHON_MODULE(pySS)
{
    def("listPidsOnHost", &pFrameWork::listPidsOnHost);
    
    class_<pHost>("Host", init<std::string>())
        .def("hostname", &pHost::hostname)
        .def("describe", &pHost::describe)
        .def("__repr__", &pHost::repr)
    ;
    
    class_<pFile>("File", init<std::string, pHost>())
        .def("filename", &pFile::filename)
        .def("hostname", &pFile::hostname)
        .def("host", &pFile::getHost, return_internal_reference<>())
        .def("describe", &pFile::describe)
        .def("__repr__", &pFile::repr)
    ;
    
    class_<pProcess>( "Process", init< unsigned long, pHost>() )
        .def("listThreads", &pProcess::listThreads)
        .def("host", &pThread::getHost,   return_internal_reference<>())
        .def("file", &pThread::getFile, return_internal_reference<>())
        .def("pid",  &pThread::getPid)
        
        // Description and overloaded python methods
        .def("describe", &pProcess::describe)
        .def("__repr__", &pProcess::repr)
    ;
    
    class_<pRank>( "Rank", init< pFile, unsigned long>() )
        .def("host", &pThread::getHost,   return_internal_reference<>())
        .def("file", &pThread::getFile, return_internal_reference<>())
        .def("rank",  &pThread::getRank)

        // Description and overloaded python methods
        .def("describe", &pRank::describe)
        .def("__repr__", &pRank::repr)
    ;
    
    class_<pPosixThread>( "PosixThread", 
        init< unsigned long, unsigned long, pHost>() )
        .def("host", &pThread::getHost,   return_internal_reference<>())
        .def("file", &pThread::getFile, return_internal_reference<>())
        .def("pid", &pThread::getPid)
        .def("tid", &pThread::getPosixThreadId)
        .def("describe", &pPosixThread::describe)
        .def("__repr__", &pPosixThread::repr)
    ;
    
    class_<pExperiment>( "Experiment", init<pFile, std::string>())
        .def(init<std::string>())
        .def("file", &pExperiment::getFile, return_internal_reference<>())

        // expAtatch command
        .def("attachToProcess", &pExperiment::attachToProcess)
        .def("attachToPosixThread", &pExperiment::attachToPosixThread)
        .def("attachToMPIRank", &pExperiment::attachToMPIRank)

        // expDetach command
        .def("detachProcess", &pExperiment::detachProcess)
        .def("detachMPIRank", &pExperiment::detachMPIRank)
        .def("detachPosixThread", &pExperiment::detachPosixThread)

        .def("detachAll",  &pExperiment::detachPosixThread)

        // expClose command
        .def("close", &pExperiment::close )
        
        // expContinue command
        .def("continue", &pExperiment::restart)

        // expStop command
        .def("stop", &pExperiment::suspend)

        // expSaveFile command
        .def("saveFile", &pExperiment::saveToFile)
        
        // expSetParam command
        .def("setParam", &pExperiment::setParam)
        
        // expView
        .def("extractResults", &pExperiment::extractResults)

        // Description and overloaded python methods
        .def("describe", &pExperiment::describe)
        .def("__repr__", &pExperiment::repr)
    ;
    
    class_<pFrameWork>("FrameWork", init<>())
        .def("listDefinedExperiments", &pFrameWork::listDefinedExperiments)    

        .def("listHostsInCluster", &pFrameWork::listHostsInCluster)    
    ;

//    to_python_converter< pPosixThread, describe<pPosixThread> >();
//    to_python_converter< pProcess, describe<pProcess> > ();
}

/* Python code
import pySS
h=pySS.Host("localhost")
f=pySS.File("./a.out", h)
h.describe()
f.describe()
exp=pySS.Experiment(pySS.File("./a.out",pySS.Host("localhost")),"pcsamp")
exp.describe()
p=pySS.Process(1234, pySS.Host("localhost"))
rid=pySS.Rank( pySS.File("./a.out", pySS.Host("localhost")), 1)
exp.attachToMPIRank( rid )
l=p.listThreads()
exp.attachToPosixThread( l[0] )
*/

#endif

File: pcli/pyapi/pyapi.py
import OSSParser
import pySS

def rptidargs( args ):
    if args.has_key('host'):
        host=pySS.Host(args['host'])
    else:
        host=pySS.Host('localhost')
    # pid
    if args.has_key('pid'):
        # and a thread
        if args.has_key('tid'):
            return pySS.PosixThread(args['pid'], args['tid'], host)
        else:
            return pySS.Process( args['pid'], host)
    # a tid, but no pid (because of previous case)
    if args.has_key('tid'):
        # FIXME: We should have a knowledge of the focus
        return pySS.PosixThread(-1, args['tid'], host)
    #
    # rank
    if args.has_key('rank'):
        if args['rank'].has_key('file'):
            file=pySS.File(args['rank']['file'], host )
        else:
            file=pySS.File('<unknwon_file>', host )
        if args['rank'].has_key('rid'):
            return pySS.Rank( file, args['rank']['rid'] )
    #
    #
    # There's an error if we get up to there
    return None


def expAttachCmd( cmd ):
    exp = pySS.Experiment("just a try")
    if cmd['expAttach'].has_key( 'rptid'):
        thread=rptidargs( cmd['expAttach']['rptid'] )
        # print "type(thread)=", type(thread)
        if type(thread) is pySS.PosixThread:
            exp.attachToPosixThread( thread )
        else:
            if type(thread) is pySS.Process:
                exp.attachToProcess( thread )
            else:
                if type(thread) is pySS.Rank:
                    exp.attachToMPIRank( thread )
                else:
                    print "Unknwon type for thread: ", type( thread )


def expCreateCmd( cmd ):
    args=cmd['expCreate']
    # 
    if args.has_key('host'):
        host=pySS.Host(args['host'])
    else:
        host=pySS.Host('localhost')
    #
    if args.has_key( 'file'):
        file=pySS.File( cmd['expCreate']['file'], host )
    else:
        file=pySS.File( "<unknwon file>", host )
    #
    if args.has_key( 'expType' ):
        exp=pySS.Experiment( file, args['expType'])
        exp.describe()



def listPidsCmd( cmd ):
    args=cmd['listPids']
    host=pySS.Host( 'localhost')
    if type(args) is dict:
        if args.has_key('host'):
            host=pySS.Host( args['host'])
    lpids= pySS.listPidsOnHost( host )
    print lpids

def semantic( cmd ):
    if cmd.has_key('listPids'):
        listPidsCmd( cmd )
    if cmd.has_key('expAttach'):
        expAttachCmd( cmd )
    if cmd.has_key('expCreate'):
        expCreateCmd( cmd )


